### DISCUSSION TITLE
[v7.39.4] How to have correct type for "fields" from useFieldArray in another file

### DISCUSSION THREAD

**Discussion by djowvani** [react-hook-form/react-hook-form#9415] [Q&A] [answered]
I'd like to have my `fields` from `useFieldArray` come with the correct type, that should look something like this `FieldArrayWithId<{...}, "users", "id">[]` so I can map those and access properly their `name` and `age` properties.

Problem is, I can't seem to make it work when doing so in another file, the type for `fields` always come as `Record<"id", string>[]`.

So far I've tried:
- Using the generic type in `useForm` on the other file
- Forcing type assertion for `fields` using the correct type
- Even though the doc says it's optional, passing `control` from my `useFormContext` to the `useFieldArray`

Any help is welcome, thank you so much!

Edit: here is a sandbox environment with this issue and comments about it https://codesandbox.io/s/react-hook-form-usefieldarray-rules-forked-jw970u?file=/src/index.js

**Comment by getTobiasNielsen** [comment, marked as answer]
I propose using `control` so the other useFieldArray knows what you want it to represent.

**Reply by getTobiasNielsen** [reply]
You worte: "Even though the doc says it's optional, passing control from my useFormContext to the useFieldArray"
I think you're making a mistake if it doesn't work, if you send a cbs, we could maybe see where.

**Comment by Moshyfawn** [comment]
With TS, you can export the custom defined `FormValues` type to import in your useFieldArray component file and pass it as the first generic. With plain JS, you can only pass `control` to `useFieldArray` as a prop, as @getTobiasNielsen mentioned above

**Comment by wpp-roman-gryndii** [comment]
```
export interface YourFormValues extends FieldValues {
  attachments: {}[]
}
form = useForm<YourFormValues>(

useFieldArray<YourFormValues, 'attachments', 'formId'>
```
